contracts/Payroll.sol

```diff
@@ -223,29 +260,46 @@ contract Payroll is Initializable, OwnableUpgradeable, ReentrancyGuardUpgradeabl
             // return the leftover of _erc20TokenOrigin
             TransferHelper.safeTransfer(_erc20TokenOrigin, msg.sender, leftOver);
         }
-        return totalAmountIn;
     }
 
-    function toAddress(bytes memory _bytes, uint256 _start) internal pure returns (address) {
-        require(_start + 20 >= _start, "toAddress_overflow");
-        require(_bytes.length >= _start + 20, "toAddress_outOfBounds");
-        address tempAddress;
+    function _performSwapV3ETH(
+        uint256 _totalAmountToSwap,
+        uint32 _deadline,
+        SwapV3[] calldata _swaps
+    ) internal returns (uint256) {
+        require(msg.value >= _totalAmountToSwap, "Payroll: Not enough msg.value");
+        address weth = IUniswapV3(swapRouter).WETH9();
 
-        assembly {
-            tempAddress := div(mload(add(add(_bytes, 0x20), _start)), 0x1000000000000000000000000)
+        uint256 totalAmountIn = 0;
+        for (uint256 i = 0; i < _swaps.length; i++) {
+            require(_swaps[i].path.length > 0, "Payroll: Empty path");
```

Comment:
I would add also the same check as in the V2 version `require(_swaps[i].path[0] == weth, "Payroll: Swap not native token");`
In this case instead of path[0] we should get the first 20 bytes for the address tough

Reply:
In v3 the token origin is at the **last** 20 bytes
`getPath(tokenA.address, poolFee, weth.address)`
this is how it looks like
`require(_swaps[i].path.toAddress(_swaps[i].path.length - 20) == weth, "Payroll: Swap not native token");`